### ISSUE TITLE
Inconsistent error message when eng.traineddata not found

### ISSUE THREAD

**Issue by teo1978** [tesseract-ocr/tesseract#221]
```
# TESSDATA_PREFIX=/usr/share/tesseract-ocr/tessdata

# echo $TESSDATA_PREFIX
/usr/share/tesseract-ocr/tessdata

# tesseract test.jpg test.txt digits
Tesseract Open Source OCR Engine v3.04.00 with Leptonica
Error opening data file /usr/share/tessdata/eng.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to the parent directory of your "tessdata" directory.
Failed loading language 'eng'
Tesseract couldn't load any languages!
Could not initialize tesseract.
```

Note: there is indeed no `eng.traineddata` file in `/usr/share/tesseract-ocr/tessdata`  (there are a bunch of other `eng.xxx` files but no .traineddata), so it is expected to get an error, ***but*** the error message says it failed to open ***/usr/share/tessdata/eng.traineddata*** while `TESSDATA_PREFIX` is set to ***/usr/share/tesseract-ocr/tessdata***.
So, either the file path in the error message is not the actual path of the file that is not found, or tesseract is ignoring the variable TESSDATA_PREFIX, in which case the error message is wrong when it says "make sure the TESSDATA_PREFIX environment variable....".

Whichever the case, something is buggy, regardless of the fact that I have no eng.traineddata file anywhere and I don't expect tesseract to work. I just expect a consistent error message.

**Comment by amitdo** [comment]
> Please make sure the TESSDATA_PREFIX environment variable is set to the **parent directory** of your "tessdata" directory.

So, if your `tessdata` dir is in the `/usr/share/tesseract-ocr` dir, `TESSDATA_PREFIX` should be set to `/usr/share/tesseract-ocr`.

**Comment by amitdo** [comment]
Also, as @stwell said elsewhere: 
> Setting TESSDATA_PREFIX is not needed as long as your tessdata directory is at the right place ($PREFIX/share/tessdata).

**Comment by amitdo** [comment]
Also read this:
http://stackoverflow.com/a/7328289

**Comment by teo1978** [comment]
> So, if your tessdata dir is in the /usr/share/tesseract-ocr dir, 
> TESSDATA_PREFIX should be set to /usr/share/tesseract-ocr.

Oh, that's strange, because I later figured out the `export` thing, and I did `export TESSDATA_PREFIX=/usr/share/tesseract-ocr/tessdata`, and it did work. Actually, it works both with and without `/tessdata`

Now I get `read_params_file: Can't open digits`, but that's another story I guess.

> Setting TESSDATA_PREFIX is not needed as long as your tessdata directory is at the right place
> ($PREFIX/share/tessdata).

Of course, but it obviously isn't in my case.

**Comment by zdenop** [comment]
@teo1978: well it could be the same story, because config files has to be with expected directory structure in TESSDATA_PREFIX...

**Comment by teo1978** [comment]
> @teo1978: well it could be the same story, because config files has to be with expected directory 
> structure in TESSDATA_PREFIX...

No, because I fixed `export TESSDATA_PREFIX=/usr/share/tesseract-ocr`, and the *Can't open digits* error persists

**Comment by zdenop** [comment]
... so digits file should be in $TESSDATA_PREFIX/tessdata/configs

**Comment by teo1978** [comment]
It wasn't. Apparently the installation of the language package was a mess. Some files (including configs/digits) were in /usr/share/tessdata; others (eng.* but not eng.traineddata) were in /usr/share/tesseract-ocr/tessdata; and eng.traineddata wasn't anywhere (I'm positive because I did a `find`), so I had downloaded it manually where the other eng.* files were...

**Comment by jbreiden** [comment]
This report appears to be from a Debian system or similar. If Tesseract was installed via the standard mechanism (apt-get) then the packaging maintainer has already made life easy. Because everything goes to a standard and known location, there is no need to set any environment variables such as TESSDATA_PREFIX. Or take any other manual installation action. The English language data (eng.traineddata) is in the tesseract-ocr-eng package and I'm not sure why you didn't see it. I am the Tesseract package maintainer for Debian systems.

```
$ dpkg -L tesseract-ocr-eng
/usr/share/tesseract-ocr/tessdata/eng.traineddata
...
```

**Comment by teo1978** [comment]
Yes this is on Debian 6.

>  If Tesseract was installed via the standard mechanism (apt-get) then the package maintainer has already made life easy.

Or he should have. We had installed tesseract-ocr and tesseract-ocr-eng via `apt-get`, and it was so broken (or so obsolete) that `tesseract -v` would issue an error message saying `-v` was an invalid option. That's why I went through the hell of compiling it from scratch.

> The English language data (including eng.traineddata) is in the tesseract-ocr-eng package.

Maybe the eng.traineddata file was not included, for whatever reason, in the Debian 6 version of the package. I have seen a lot of questions in forums about the same issue (eng.traineddata missing after installing the package).

**Comment by jbreiden** [comment]
You can think of Debian as a curated software collection. Debian 6 was released on February 6th, 2011 and all its packages are from that era. It contained Tesseract 2.04 which is ancient history. Please note that Debian 6 is officially obsolete, and the current stable release is Debian 8. 
See https://www.debian.org/releases/

Unlike some other operating systems, best practice for Debian is to update the entire distribution if you want more recent versions of programs. Trying to install a modern Tesseract on an older Linux distribution is really tricky, especially because there are a lot of dependencies involved. I strongly do not recommend it.

**Comment by teo1978** [comment]
Actually debian 6 is LTS until this 29 of February https://wiki.debian.org/LTS

So, I would have expected a packgage for Debian 6 to be maintained to not become "ancient history" until, well, just now.

**Comment by stweil** [comment]
You missed several important points:

* Debian LTS only provides security maintenance.
* Debian LTS is not supported by the normal Debian maintainers, but by a separate group of volunteers and companies.

As @jbreiden already said, you have the option to upgrade to a newer version of Debian. Or you can also take any other current distribution.

**Comment by 1manfactory** [comment]
Can some one PLEASE just provide a solution to this error message WITHOUT discussing pros and cons back to the stone age?

**Comment by stweil** [comment]
Sure. The solution is already there: install Tesseract on a recent Linux distribution.

**Comment by khanglam** [comment]
After days of searching for the solution, none of the proposed solutions worked for me because I am using objective C++ in xcode. But after tons of experimentations, for anyone that still need this solved, the solution is a 1-liner (if ure using TessBaseAPI), before api.init(...) add 
G8Tesseract *tesseract = [[G8Tesseract alloc] initWithLanguage:@"eng"];
This magically gets rid of the TESSDATA_PREFIX error

**Comment by miramind** [comment]
The solution is here:
change your tessdata preffix to TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata and all will be good. It works for fixing 4.0.16 version

**Comment by gautamkumarjaiswal** [comment]
@miramind Thanks, it worked for me.

**Comment by ubedthaheem** [comment]
Hello, i am facing same error while using OCR , here is error which i got from debug mode in VB 2019.
`rror opening data file E:\MBHamzOCR_New\bin\Debug/Sindhi.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Failed loading language 'Sindhi'
Tesseract couldn't load any languages!
Exception thrown: 'Tesseract.TesseractException' in Tesseract.dll
The program '[17276] MB Hamz Ali Sindhi OCR.exe' has exited with code 0 (0x0).`

**Comment by ubedthaheem** [comment]
Hello, i am facing same error while using OCR , here is error which i got from debug mode in VB 2019.
`rror opening data file E:\MBHamzOCR_New\bin\Debug/Sindhi.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Failed loading language 'Sindhi'
Tesseract couldn't load any languages!
Exception thrown: 'Tesseract.TesseractException' in Tesseract.dll.`


i am using 3 traineddata files including Sindh, Urdu and English, but its not working for me.

**Comment by Shreeshrii** [comment]
Check first by using tesseract from command line

Test with each language separately

Set tessdata_prefix

**Comment by ubedthaheem** [comment]
where to set tessdata_prefix?? i m worried about it

**Comment by zdenop** [comment]
You should set it as environment variable of your operation system.

**Comment by ubedthaheem** [comment]
i set environment variable but error remains same.. 
`Error opening data file E:\OCR_FILES\bin\Debug/snd.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Failed loading language 'snd'
Tesseract couldn't load any languages!
`
* snd = Sindhi language trained data file.

**Comment by Shreeshrii** [comment]
You have to make sure that file exists in the path specified by you. You will have to modify following commands for windows (use dir instead of ls -l)

```
ls -l ~/tessdata_best/snd.traineddata 
-rw-rw-r-- 1 ubuntu ubuntu 11981538 Jul 22 09:27 /home/ubuntu/tessdata_best/snd.traineddata


tesseract sindhi-test.png - -l snd --psm 6 --oem 1 --tessdata-dir ~/tessdata_best

```
 <div dir="rtl">
وڪيبيڊبا تي ڀليڪار»<br>
کليل ڄاڻ ڇيڪلو جنهن کي ڪير به سنواري سگهي ٿو<br>
سنڌي ۾ 11,864 مرهمرون آهن.<br>
</div>

![sindhi-test](https://user-images.githubusercontent.com/5095331/62409555-e9c38d00-b5f6-11e9-95f1-2931b2af1e5d.png)

**Comment by ubedthaheem** [comment]
i am using open source VietOCR on VB. yes traineddata file exists and i set environment variable too... but same problem.

**Comment by Shreeshrii** [comment]
@nguyen Can you suggest solution?

**Comment by ubedthaheem** [comment]
actually, program is not reading 'tessdata' folder, i moved data file into root folder 'Debug' it worked, but an other error occurred, that is  `read_params_file: Can't open tess_configs
Exception thrown: 'System.IndexOutOfRangeException'` which means program is not reading tessdata directory folder. 
note that, i set system environment variable to : TESSDATA_PREFIX=E:\MBHamzOCR_New\bin\Debug\tessdata\
need help to fix this issue.. :)

**Comment by nellyonlinux** [comment]
I had the same problem, then I realised that the problem was actually rather simple and a fix was right there - the error message is explicit about where the files are expected to be and that is in the **parent directory** of tessdata.

_Please make sure the TESSDATA_PREFIX environment variable is set to the parent directory of your "tessdata" directory_

So it seems a configuration file expects the files it needs to be one level up from the location where they were installed: 

The solution is to copy the language files and the training data (in my case eng.traineddata and osd.traineddata) from the tessdata folder /usr/share/tesseract-ocr/4.00/tessdata/  to the parent folder one level up /usr/share/tesseract-ocr/4.00/

Starting from the required files' original location I copied them to the parent folder like this:

`sudo cp -i eng.traineddata osd.traineddata /usr/share/tesseract-ocr/4.00/`

After this tesseract did not have any more problems.

**Comment by stweil** [comment]
@nellyonlinux, that is not the right way to do things. If your Tesseract installation for Linux has a `/usr/share/tesseract-ocr/4.00/tessdata`, then traineddata files should be in that directory. And don't set `TESSDATA_PREFIX`! Tesseract normally knows where to find the data files without that hack.

**Comment by nellyonlinux** [comment]
I am willing to learn but I found no way to set TESSDATA_PREFIX on various forums. It would appear that this issue has cropped up several times over the years with no clear solution for less technical types like myself. 

The files required are in that directory but some other settings point elsewhere - I leaned that from the error message that tesseract expected the files to be in the parent folder, so copied them over and it worked. You are probably right that it should not be fixed like that but it was a solution that worked for me.
